### PR TITLE
Expanded vec128 and VectorSub

### PR DESCRIPTION
**Pull request by HaydnTrigg**
Added double precision (f64) values to the vec128 structure
Added cases for all formats of data for vector subtraction in the
Value::VectorSub function

NOTE: Unsure what the saturate function is for however maintained
original functionality

### CODE REVIEW

**Comment by Razzile** [comment]
I reckon ben will ask you to run `xb format` to format the code.

**Comment by DrChat** [comment]
Neat - this is cool. Like @Razzile said, just format your code according to [this guide](https://github.com/benvanik/xenia/blob/master/docs/style_guide.md) and we'll pull it!

File: src/xenia/cpu/hir/value.cc [hunk 1]
```diff
@@ -1004,29 +1004,80 @@ void Value::VectorRol(Value* other, TypeName type) {
...
-      }
-    default:
-      assert_unhandled_case(type);
-      break;
-  }
+	bool saturate) {
+	assert_true(this->type == VEC128_TYPE && other->type == VEC128_TYPE);
+	switch (type) {
+	case FLOAT32_TYPE:
+		if (saturate) assert_always();
```
Comment (DrChat) [inline src/xenia/cpu/hir/value.cc]: Surround if statements in curly brackets please

**Comment by ScorchEmber256** [comment]
This only fixes the wonky physics in sonic games, however the death issues and inverted model glitches are still apparent. Likely caused by #514. I think that commit was responsible for invisible level in sonic unleashed demo, and model depth issues in TLoS:DotD and various other games.

xenia spv_translator_f8d9472_may 27 2016 - legend of spyro 29_05_2016 10_49_48 am
xenia spv_translator_f8d9472_may 27 2016 - legend of spyro 29_05_2016 10_49_37 am

**Comment by DrChat** [comment]
I've actually got some uncommitted code locally that implements just about everything that was unimplemented here.
Will get around to committing it when I'm wrapping up the Vulkan branch.

**Comment by ScorchEmber256** [comment]
*Insert flattery statement, because your amazing!

**Comment by HaydnTrigg** [comment]
I've updated the code with Google formatting, for some reason `xb format` doesn't work for formatting any files it just throws up a fuss about a file either not being committed or once committed no changes (self defeating problem).

PS. I'm going to write all of the SSE and AVX acceleration for this shortly, @DrChat should I wait until you've put up your latest file?

File: src/xenia/cpu/hir/value.cc [hunk 2]
```diff
@@ -1,11 +1,11 @@
...
- * Released under the BSD license - see LICENSE in the root for more details. *
- ******************************************************************************
- */
+******************************************************************************
+* Xenia : Xbox 360 Emulator Research Project                                 *
+******************************************************************************
+* Copyright 2013 Ben Vanik. All rights reserved.                             *
+* Released under the BSD license - see LICENSE in the root for more details. *
+******************************************************************************
+*/
```
Comment (DrChat) [inline src/xenia/cpu/hir/value.cc]: The space needs to come back - just undo this change.

File: src/xenia/cpu/hir/value.cc [hunk 3]
```diff
@@ -1003,81 +1003,81 @@ void Value::VectorRol(Value* other, TypeName type) {
...
-		assert_unhandled_case(type);
-		break;
-	}
+void Value::VectorSub(Value* other, TypeName type, bool is_unsigned,
+                      bool saturate) {
+  assert_true(this->type == VEC128_TYPE && other->type == VEC128_TYPE);
+  switch (type) {
+    case FLOAT32_TYPE:
+      if (saturate)
+        assert_always();
```
Comment (DrChat) [inline src/xenia/cpu/hir/value.cc]: Wrap these one-liner if statements in brackets and you'll be golden!

**Comment by DrChat** [comment]
To be honest I'd recommend not writing any AVX/SSE/etc acceleration for this code - it already runs fast enough.
If you want to meddle with some x64 assembly, check out x64_backend.cc and take a stab at natively implementing any instructions that are being emulated with a function call back to C++. IIRC, there should still be a few left.

Also - If you already know how, could you squash the 2 formatting commits back to one?

File: src/xenia/base/vec128.h [hunk 4]
```diff
@@ -90,7 +90,8 @@ typedef struct alignas(16) vec128_s {
       uint32_t uz;
       uint32_t uw;
     };
-    float f32[4];
+	float f32[4];
+	double f64[2];
```
Comment (DrChat) [inline src/xenia/base/vec128.h]: Format this too please

**Comment by HaydnTrigg** [comment]
@ScorchEmber256 The code was originally written against missing Halo 3 and Halo Reach functionality with the processing of entity positions, I can only assume by the code that I have written that I've just added missing functionality I'm not entirely sure how it would affect other games.

File: src/xenia/base/vec128.h (hunk 4 above)
Comment (HaydnTrigg) [inline src/xenia/base/vec128.h]: I thought the formatting for this file was okay?

Comment (DrChat) [inline src/xenia/base/vec128.h]: Should be spaces instead of tabs

Comment (HaydnTrigg) [inline src/xenia/base/vec128.h]: Fixed

Comment (HaydnTrigg) [inline src/xenia/base/vec128.h]: On the SSE and AVX stuff. I noticed Halo Reach and Halo 3 use this code a lot and when I debugged it appeared to be positional co-ordinates, thats why I was eager to smash out some faster code for it

**Comment by DrChat** [comment]
@HaydnTrigg This code's part of the recompiler - this is ran once during compilation to propagate constants through the generated code. The speed on this code has no bearing on the speed of the generated code.
I understand your eagerness though - just keep looking for areas to improve! There's plenty of stuff that needs it.

**Comment by DrChat** [comment]
Merged in 7fe4ffcaee77c4fe1705c9c8e9176fa17cc09790.
